### PR TITLE
[WIP] Add greedynas configs

### PR DESCRIPTION
**Pull request by humu789**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        dict(type='ColorTransform', magnitude=0.4, prob=0.6),
+        dict(type='Contrast', magnitude=0.8, prob=1.)
+    ],
+    [dict(type='Equalize', prob=0.8),
+     dict(type='Equalize', prob=0.6)],
+]
+
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='RandomResizedCrop', size=224),
```
Comment (pppppM) [inline configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py:105]: add `backend='pillow'`

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    dict(type='Resize', size=(256, -1)),
+    dict(type='CenterCrop', crop_size=224),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='ImageToTensor', keys=['img']),
+    dict(type='ToTensor', keys=['gt_label']),
+    dict(type='Collect', keys=['img', 'gt_label'])
+]
+test_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='Resize', size=(256, -1)),
```
Comment (pppppM) [inline configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py:126]: add `backend='pillow'`

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# dataset settings
+dataset_type = 'ImageNet'
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='RandomResizedCrop', size=224),
```
Comment (pppppM) [inline configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py:8]: add `backend='pillow'`

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
+    dict(type='ColorJitter', brightness=0.0, contrast=0.0, saturation=0.0),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='ImageToTensor', keys=['img']),
+    dict(type='ToTensor', keys=['gt_label']),
+    dict(type='Collect', keys=['img', 'gt_label'])
+]
+train_val_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='Resize', size=(256, -1)),
```
Comment (pppppM) [inline configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py:18]: add `backend='pillow'`

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    dict(type='CenterCrop', crop_size=224),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='ImageToTensor', keys=['img']),
+    dict(type='ToTensor', keys=['gt_label']),
+    dict(type='Collect', keys=['img', 'gt_label'])
+]
+test_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='Resize', size=(256, -1)),
```
Comment (pppppM) [inline configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_greedynas.py:27]: add `backend='pillow'`

**Review by pppppM — CHANGES_REQUESTED** [review]

File: configs/_base_/schedules/mmcls/imagenet_bs1024_greedynas.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# optimizer
+optimizer = dict(type='SGD', lr=0.12, momentum=0.9, weight_decay=4e-5)
```
Comment (wutongshenqiu) [inline configs/_base_/schedules/mmcls/imagenet_bs1024_greedynas.py:2]: As mentioned in subsection 5.1 of the paper, when training supernet, they use SGD with **Nesterov acceleration**, but that option does not seem to be set here.

File: configs/nas/greedynas/greedynas_supernet_mobilenet_8xb128_in1k.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+norm_cfg = dict(type='BN')
+model = dict(
+    type='mmcls.ImageClassifier',
+    backbone=dict(
+        type='SearchableMobileNet',
+        arch_setting_type='greedynas',
+        widen_factor=1.0,
+        norm_cfg=norm_cfg,
+        act_cfg=dict(type='ReLU')),
```
Comment (wutongshenqiu) [inline configs/nas/greedynas/greedynas_supernet_mobilenet_8xb128_in1k.py:15]: I notice that all activation layers in searchable block used `HSwish`. However, `act_cfg` here used `ReLU`, which will affect the first and last `ConvModule` in mobilenet.
Is that expected?

File: configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py (hunk 1 above)
Comment (humu789) [reply configs/_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py:105]: OK

File: configs/nas/greedynas/greedynas_supernet_mobilenet_8xb128_in1k.py (hunk 7 above)
Comment (humu789) [reply configs/nas/greedynas/greedynas_supernet_mobilenet_8xb128_in1k.py:15]: yes, it is confirmed with huangtao

File: mmrazor/core/optimizer/rmsproptf.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                 momentum=0.,
+                 centered=False,
+                 decoupled_decay=False,
+                 lr_in_momentum=True):
+        if not 0.0 <= lr:
+            raise ValueError('Invalid learning rate: {}'.format(lr))
+        if not 0.0 <= eps:
+            raise ValueError('Invalid epsilon value: {}'.format(eps))
+        if not 0.0 <= momentum:
+            raise ValueError('Invalid momentum value: {}'.format(momentum))
```
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/core/optimizer/rmsproptf.py:53]: This directly ref to huangtao's code, but yours is better

File: configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
+_base_ = [
+    '../../_base_/datasets/mmcls/imagenet_bs128_colorjittor_aa_greedynas.py',
+    '../../_base_/schedules/mmcls/imagenet_bs1024_aa_greedynas.py',
+    '../../_base_/mmcls_runtime.py'
+]
+
+data = dict(samples_per_gpu=96)
+optimizer = dict(lr=0.048)
+lr_config = dict(
+    step=int(1666 * 2.4), warmup_iters=1666 * 3, warmup_ratio=1e-6 / 0.048)
```
Comment (wutongshenqiu) [inline configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py:10]: Does `1666` here indicate the number of iterations per epoch?

But according to the instruction [here](https://image-net.org/download.php), the number of total training images is `1281167`, and if batch size is `96 * 8 = 768`, the iterations of one epoch might be `ceil(1281167 / 768) = 1669` since `drop_last` is set to `True`

File: configs/_base_/schedules/mmcls/imagenet_bs1024_aa_greedynas.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# optimizer
+optimizer = dict(
+    type='RMSpropTF', lr=0.064, eps=0.001, weight_decay=1e-5, momentum=0.9)
+optimizer_config = dict(grad_clip=None)
```
Comment (pppppM) [inline configs/_base_/schedules/mmcls/imagenet_bs1024_aa_greedynas.py:4]: It seems should be `grad_clip=dict(max_norm=5, norm_type=2)`

File: configs/_base_/schedules/mmcls/imagenet_bs1024_aa_greedynas.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# optimizer
+optimizer = dict(
+    type='RMSpropTF', lr=0.064, eps=0.001, weight_decay=1e-5, momentum=0.9)
```
[the remark this answers is not in the document]
Comment (pppppM) [reply configs/_base_/schedules/mmcls/imagenet_bs1024_aa_greedynas.py:3]: https://github.com/open-mmlab/mmrazor/blob/e4e95131a2451f748923fe702c269d02c9a753f6/configs/_base_/schedules/mmcls/imagenet_bs1024_spos.py#L2

File: configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    placeholder_mapping=dict(
+        searchable_blocks=dict(
+            type='OneShotOP',
+            choices=dict(
+                mb_k3e3=dict(
+                    type='MBBlock',
+                    kernel_size=3,
+                    expand_ratio=3,
+                    norm_cfg=norm_cfg,
+                    drop_path_rate=0.2,
```
Comment (pppppM) [inline configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py:54]: In huangtao's code, seems `drop_path_rate=0`.
He seems just uses a `DropOut` before the FC layer, and the drop prob is 0.2.

File: configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py (hunk 9 above)
Comment (humu789) [reply configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py:10]: Yes, but training dataset is splited for trianing and evaluating,  (1281167 - 1024) // 768 = 1666

Comment (wutongshenqiu) [reply configs/nas/greedynas/greedynas_subnet_mobilenet_8xb96_in1k.py:10]: got it.
